templates/README.md

```diff
@@ -0,0 +1,276 @@
+    # sata, virtio, ...)
+    defaults.template.cnv.io/disk: default-disk
+
+    # The goal of default volume is to be able to configure mostly
+    # performance parameters like caches if those are exposed
+    # by the underlying volume implementation.
+    defaults.template.cnv.io/volume: default-volume
+
+    # The goal of default network is similar to default-disk
+    # and should be used as a template to ensure OS compatibility
+    # and performance
+    defaults.template.cnv.io/nic: default-nic
+
+    # The goal of default network is similar to default-volume
+    # and should be used as a template that specifies performance
+    # and connection parameters (L2 bridge for example)
+    defaults.template.cnv.io/network: default-network
+
+    # Replacements useful for finishing the transformation
+    # This is an extension we could use for specifying common
+    # fields for all disks/nics. Requires extra implementation
+    # that will do the find and replace using a syntax documented
+    # here: https://kubernetes.io/docs/reference/kubectl/jsonpath/
+    # The jsonpath root is the objects: element of the template
+    # XXX There is currently an issue with annotations not supporting
+    #     list values
+    template.cnv.io/replace:
+    - path: "/VM/disks/*/type"
+      value: virtio
+
+    # Another extension useful for CNV aware tooling that allows
+    # expressing additional validation rules for this template
+    # The jsonpath root is the objects: element of the template
+    # XXX There is currently an issue with annotations not supporting
+    #     list values
+    template.cnv.io/validate:
+    - path: "/VM/spec/domain/guest/memory"
+      min: 3
+      max: 4
+
+    # Another extension for hinting at which elements should be
+    # considered editable. The content is a line separated
+    # list of jsonpath selectors.
+    # The jsonpath root is the objects: element of the template
+    template.cnv.io/editable: |
+      /objects[0].spec.template.spec.domain.cpu.cores
+      /objects[0].spec.template.spec.domain.resources.requests.memory
+      /objects[0].spec.template.spec.domain.devices.disks
+      /objects[0].spec.template.spec.volumes
+      /objects[0].spec.template.spec.networks
+
+  labels:
+    # The UI can show all possible template.cnv.io/* values in a nice way
+    # and let the user filter down the available templates to the one
+    # the user actually wants:
+    # A single selected template only means no conflicts and no smart
+    # merging code. This has to be done using labels to allow efficient
+    # searching.
+    # The format has the following meaning:
+    # {os,flavor,size}.template.cnv.io/{value}: true (or false for exclusion)
+    # OS names should match the libosinfo identifiers
+    # flavors are tiny, medium, large, etc.
+    # workloads are generic, high-performance, io-intensive,
+    #               oracle-db, sap-hana...
+    os.template.cnv.io/windows2k12r2: "true"
```

Comment:
The scheme does not work.

It works to use boolean to flag a template if it is relevant for an os, workload, size.

What does not work is that the triple is always depending on each other, thus we need to say, this template is applicable to "large windows compute", but from the current mechanism it is not clear if this template applies to "large windows io-intensive" and "large windows minilal" or just one of them.

Instead we could encode the triples in flags:

```
guest.cnv.io/large-windows7-compute
guest.cnv.io/large-windows7-generic
```

Reply:
(os1 OR os2 OR os3) AND (workload1 OR workload2) AND (flavor1 OR flavor2) AND NOT anything that is false

And I am not going to express the full boolean logic there as I do not see the need for it. We can use the proposed negative (false) for exceptions and the rest will be implemented using separate template files.

Reply:
Could you please explain where you would put this logic?

Reply:
Label based object search is already implemented in kubernetes. And searching for a set of labels will be done by the user via the UI. The fact we will organize the templates to follow the described logic does not require any implementation. It will just be done that way when we ship the templates.

Reply:
Rigth. I understand that in Kubernets you can do boolean expressions.

My point is: By havnig the labels for size, os, and workload independently on a VM, we don#t knwo what triple is support.
This means that all possible permutations of all labels on a VM are supported.
If this is th goal then good, but to me it looks greedy.

Reply:
Yes it is greedy by design.